### PR TITLE
LockFreeLinkedSet: add initial implementation

### PR DESCRIPTION
**Pull request by ianks**
This PR adds a lock-free set, implemented as a link list. This data-structure is neccesary as a building block for a lock-free concurrent hash table/cache. The only difference is that in a concurrent hash-table, I was use split-ordering on the nodes in the linked list instead ordering the items based on their hash value.

As always, fire away on comments and potential refactorings! Always looking to improve my code.

### CODE REVIEW

**Comment by jdantonio** [comment]
@ianks If you don't mind, I'd like to wait until the full release of 0.9 to merge this.

File: lib/concurrent/edge/lock_free_linked_set.rb [hunk 1]
```diff
@@ -0,0 +1,131 @@
+require 'concurrent/edge/lock_free_linked_set/node'
+require 'concurrent/edge/lock_free_linked_set/window'
+
+module Concurrent
+  module Edge
+    class LockFreeLinkedSet
```
Comment (pitr-ch) [inline lib/concurrent/edge/lock_free_linked_set.rb]: Could you include short description of the class and if possible a link to your source of the algorithm?

File: lib/concurrent/edge/lock_free_linked_set/node.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
+require 'concurrent/edge/atomic_markable_reference'
+
+module Concurrent
+  module Edge
+    class LockFreeLinkedSet
+      class Node < Synchronization::Object
+        include Comparable
+
+        attr_reader :Data, :SuccessorReference, :Key
```
Comment (pitr-ch) [inline lib/concurrent/edge/lock_free_linked_set/node.rb]: This does not feel right `pred.SuccessorReference.compare_and_set ...`. Could you create underscore readers for the CalmelCase ivars?

File: lib/concurrent/edge/lock_free_linked_set/window.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
+module Concurrent
+  module Edge
+    class Window
```
Comment (pitr-ch) [inline lib/concurrent/edge/lock_free_linked_set/window.rb]: Shouldn't this be inside the Set's namespace?

**Comment by pitr-ch** [comment]
And thank you for your great contributions!

File: lib/concurrent/edge/lock_free_linked_set/node.rb (hunk 2 above)
Comment (ianks) [inline lib/concurrent/edge/lock_free_linked_set/node.rb]: Yeah I can do that. TBH I don't know how I like the uppercase convention for final fields. It makes the code reading awkward IMO. I wonder if there is a more explicit way of doing it that doesn't require semantic casing. What are you thoughts?

File: lib/concurrent/edge/lock_free_linked_set/window.rb (hunk 3 above)
Comment (ianks) [inline lib/concurrent/edge/lock_free_linked_set/window.rb]: Yup, nice catch!

**Comment by ianks** [comment]
@pitr-ch @jdantonio 

I have addressed the comments on the PR. Let me know if anything else is needed! :smile:

**Pull request merged**